File: hustack/Ex15_Parentheses/codes/stack.py
def check_correct(s: str) -> int:
    stack = []
    pairs = {")": "(", "]": "[", "}": "{"}
    for char in s:
        if char in "([{":
            stack.append(char)
        elif char in ")]}":
            if not stack or stack[-1] != pairs[char]:
                return 0
            stack.pop()
    return 1 if not stack else 0

if __name__ == "__main__":
    s = input().strip()
    print(check_correct(s))

File: hustack/Ex15_Parentheses/codes/stack.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

bool check_correct(const string &s) {
    stack<char> stk;
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') {
            stk.push(c);
        } else {
            if (stk.empty()) return false;
            char top = stk.top(); stk.pop();
            if ((c == ')' && top != '(') ||
                (c == ']' && top != '[') ||
                (c == '}' && top != '{')) {
                return false;
            }
        }
    }
    return stk.empty();
}

int main() {
    string s;
    cin >> s;
    cout << (check_correct(s) ? 1 : 0) << endl;
    return 0;
}
